mautic/mautic
# Schedule campaign events at an hour relative to the contact's timezone

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | y
| New feature? | y
| Automated tests included? |y
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
This PR fixes a bug where campaign events were executed regardless of the campaign's published state.

It also adds a feature to schedule a campaign event to be executed based on a relative hour to the contact's timezone or the system configuration if none. (There was a bug where new installations did not add timezone as a custom field so this has been fixed as well.) The event can be scheduled at a specific hour, between hours, and/or specific days of the week. 

It is includes https://github.com/mautic/mautic/pull/6482 due to conflicts so please merge it first. 

[//]: # ( As applicable: )

### Published state bug
#### Steps to reproduce the bug:
1. Create a new Campaign
2. Setup Action to Send Email "at a relative time period" 5 minutes later
3. Edit the campaign and set unpublish at date/time in the past
4. Look at the database `campaign_lead_event_log` table and find the scheduled log entry for the contact and note the id
5. Execute the command `mautic:campaigns:execute --scheduled-log-ids=ID`; replace the ID with the one noted in step 4
6. Note that the event executes even though the campaign is supposed to be unpublished

#### Steps to test this PR:
1. Repeat and the event will not execute rather a log entry will note that it failed due to the campaign being unpublished (to prevent never ending attempts)

### Relative hour feature

#### Steps to test this PR:
1. Create a new campaign
2. Add an action and schedule it with a relative time
3. Notice that if minute or hour interval unit is chosen, the custom hour/DOW fields disappear. 
4. Choose 1 day and set a specific time
5. Create a contact and set the contact's timezone
6. Create a second contact and set a different timezone
7. Create a third contact and leave timezone empty
8. Add all three to the campaign and execute via the command
9. Check their timelines and the actions should be scheduled at the hour relative to their timezone. The 3rd will be scheduled based on the system default timezone.
10. Edit the campaign and remove the hour (leave it blank)
11. Add a fourth contact through the campaign, execute the command, then note that the action should be scheduled exactly 24 hours in the future
12. Wait a day, execute the command, and then make sure the actions execute and are not rescheduled
